inventory.py

```diff
@@ -93,57 +99,105 @@ def gather_course_info_for_account(account_id: int, term_id: int) -> Sequence[in
+               e.course_id AS course_id,
+               e.course_section_id AS course_section_id,
+               e.user_id AS user_id,
+               e.workflow_state AS workflow_state,
                r.base_role_type AS role_type
         FROM enrollment_dim e
         JOIN role_dim r
             ON e.role_id=r.id
-        WHERE e.course_id IN ({udw_courses_string});
+        WHERE e.course_id IN ({courses_string});
     '''
 
     logger.info('Making enrollment_dim query')
     enrollment_df = pd.read_sql(enrollment_query, UDW_CONN)
+    logger.debug(enrollment_df.head())
+    return enrollment_df
 
-    enrollment_df.to_csv(os.path.join('data', 'enrollment.csv'), index=False)
-    logger.info('Enrollment data was written to data/enrollment.csv')
 
-    user_ids = enrollment_df['enrollment_user_id'].drop_duplicates().to_list()
+def pull_user_data_from_udw(user_ids: Sequence[int]) -> pd.DataFrame:
     users_string = ','.join([str(user_id) for user_id in user_ids])
-
     user_query = f'''
-        SELECT u.id AS user_id,
-               u.canvas_id AS user_canvas_id,
-               u.name AS user_name,
-               u.workflow_state AS user_workflow_state,
-               p.unique_name AS pseudonym_uniqname
+        SELECT u.id AS warehouse_id,
+               u.canvas_id AS canvas_id,
+               u.name AS name,
+               p.unique_name AS uniqname,
+               u.workflow_state AS workflow_state
         FROM user_dim u
         JOIN pseudonym_dim p
             ON u.id=p.user_id
         WHERE u.id in ({users_string});
     '''
-
     logger.info('Making user_dim query')
     user_df = pd.read_sql(user_query, UDW_CONN)
-    user_df.to_csv(os.path.join('data', 'user.csv'), index=False)
-    logger.info('User data was written to data/user.csv')
+    # Found that the IDs are not necessarily unique, so dropping duplicates
+    user_df = user_df.drop_duplicates(subset=['warehouse_id', 'canvas_id'])
+    logger.debug(user_df.head())
+    return user_df
+
+
+def run_course_inventory() -> None:
+    start = datetime.now()
+
+    # Gather course data
+    course_df = gather_course_info_for_account(1, ENV['TERM_ID'])
+
+    # Gather enrollment data
+    udw_course_ids = course_df['warehouse_id'].to_list()
+    enrollment_df = pull_enrollment_data_from_udw(udw_course_ids)
+
+    # Gather user data
+    udw_user_ids = enrollment_df['user_id'].drop_duplicates().to_list()
+    user_df = pull_user_data_from_udw(udw_user_ids)
+
+    # Find and remove rows with nonexistent user ids from enrollment_df
+    # This can take a few minutes
+    user_ids = user_df['warehouse_id'].drop_duplicates().to_list()
+
+    def check_if_valid_user_id(id: int) -> bool:
+        if id in user_ids:
+            return True
+        else:
+            return False
+
+    enrollment_df['valid_id'] = enrollment_df['user_id'].map(check_if_valid_user_id)
+    enrollment_df = enrollment_df[(enrollment_df['valid_id'])].drop(columns=['valid_id'])
```

Comment:
This operation is necessary, but it is time intensive (a little less than three minutes, I think). I looked for alternatives like including an `INNER JOIN` in the UDW enrollment query, but I wasn't able to get anything to work. Any ideas?

Reply:
I had hoped that this would become unnecessary once we began filtering on the enrollment `workflow_state` being 'active', but I'm still getting the error, with small numbers of user IDs not being present in the `user` table.